src/lib/Formula.ts

```diff
@@ -0,0 +1,322 @@
+
+    name(): string {
+        return this.predName;
+    }
+
+    arguments(): Constant[] {
+        return this.args;
+    }
+
+    subfs(): Formula[] {
+        return [];
+    }
+
+    toString(): string {
+        const argsStr = this.args.map(arg => arg.toString()).join(',');
+        return `${this.predName}(${argsStr})`;
+    }
+
+    toTex(): string {
+        const argsStr = this.args.map(arg => arg.toTex()).join(',');
+        return `\\text{\\textsf{${this.predName}}}(${argsStr})`;
+    }
+
+    isTrue(v: Map<string, boolean | null>): boolean | null {
+        const atomString = this.toString();
+        if (v.has(atomString)) {
+            if (v.get(atomString) === null) {
+                throw new Error(`Atom "${atomString}" does not have valuation.`);
+            }
+            return v.get(atomString) as boolean;
+
+        } else {
+            throw new Error(`Atom "${atomString}" not found in the valuation.`);
+        }
+    }
+
+    atoms(): Set<PredicateAtom> {
+        return new Set([this]);
+    }
+
+    constants(): Set<string> {
+        return new Set(this.args.map(arg => arg.toString()));
+    }
+
+    predicates(): Set<string> {
+        return new Set([this.predName]);
+    }
+}
+
+
+class Negation extends Formula {
+    private original: Formula;
+
+    constructor(originalFormula: Formula) {
+        super('', '');
+        this.original = originalFormula;
+    }
+
+    public originalFormula(): Formula {
+        return this.original;
+    }
+
+    public subfs(): Formula[] {
+        return [this.original];
+    }
+  
+    public toString(): string {
+        return `-${this.original.toString()}`;
+    }
+
+    public toTex(): string {
+        return `\\lnot ${this.original.toTex()}`;
+    }
+
+    public isTrue(v: Map<string, boolean | null>): boolean | null {
+        if (!v.has(this.original.toString())) {
+            throw new Error('Atom not found in the map');
+        }
+        return !this.original.isTrue(v);
+    }
+
+    public atoms(): Set<PredicateAtom> {
+        return this.original.atoms();
+    }
+
+    public constants(): Set<string> {
+        return this.original.constants();
+    }
+
+    public predicates(): Set<string> {
+        return this.original.predicates();
+    }
+}
+
+class Conjunction extends Formula {
+    private conjuncts: Formula[];
+
+    constructor(conjuncts: Formula[]) {
+        super('&', ' \\land ');
+        this.conjuncts = conjuncts;
+    }
+
+    public subfs(): Formula[] {
+        return this.conjuncts;
+    }
+
+    public isTrue(v: Map<string, boolean | null>): boolean | null {
+        for (const form of this.conjuncts) {
+            if (!form.isTrue(v)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public atoms(): Set<PredicateAtom> {
+        const atoms = new Set<PredicateAtom>();
+
+        for (const form of this.conjuncts) {
+            form.atoms().forEach(atom => atoms.add(atom));
+        }
```

Comment:
Tu aj v ďalších podobných metódach sa hodí [`this.conjuncts.flatMap()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap), ktorý môžete použiť priamo ako argument konštruktora `Set()` a rovno výsledok vrátiť.

Navyše, podobne ako ste implementovali `toString()` už vo `Formula` využitím `subfs()`, môžete tam implementovať aj `atoms()`, `constants()`, `predicates()`.